### ISSUE TITLE
Hand-ray jittering with the Windows Mixed Reality plugin for Unity

### ISSUE THREAD

**Issue by ShayneRC** [microsoft/MixedRealityToolkit-Unity#10225]
## Describe the bug

We had to do an upgrade to Unity 2020.3.13 and MRTK 2.7.2. With that came a few hurdles that we have overcome but the main issue we are having right now is the hand-rays start jittering after the hand being active for 5 seconds. 

So I created a new unity project and added all the MRTK packages with 1 of the example samples (solver examples). Here are my findings:

•	I started by making all the project settings the same as what ours currently is: **Jittery**
•	I then re-created the project and did only the defaults needed for the WMR plugin for XR: **Jittery**
•	I noticed a project setting to do with inputs (Project Settings > Player > Other Settings > Configuration > Active Input Handling), so I built a version for each setting; one was **jittery** and the other had **no hand tracking** at all.
•	I then re-created the project and did only the defaults needed for OpenXR plugin: **Stable**

We need to use the WMR plugin due to some ARR requirements. Here is a dropbox link that shows the sample project with one being built with OpenXR and the other with WMR: https://www.dropbox.com/sh/3n0v3lpnuc7snyt/AAAi4_RYTs4bvud4ERHNgMYXa?dl=0

You might need to look closely as it’s kind of hard to notice the jittering but it is there. It’s much more noticeable on the device especially when moving the hand.

## To reproduce

1. Create a new Unity 2020.3.13 project.
2. Change the build platform to UWP.
3. Import the MRTK packages - I used the Microsoft Mixed Reality Feature Tool with these packages:
a. Mixed Reality Toolkit Examples.
b. Mixed Reality Toolkit Extensions.
c. Mixed Reality Toolkit Foundations.
d. Mixed Reality Toolkit Standard Assets.
e. Mixed Reality Toolkit Tools.
4. Using the Package Manager I imported the _Demos - Solvers_ sample from the MRTK Examples.
5. Open the sample scene - _Assets\Samples\Mixed Reality Toolkit Examples\2.7.2\Demos - Solvers\Scenes\SolverExamples.unity_ (this could be any sample, I just happen to choose this one).
6. Configure the project for MRTK - Mixed Reality > Toolkit > Utilities > Configure Project for MRTK.
7. Choosing to go the WMR path will lead to the jittery hand-ray whereas going the OpenXR path leads to a stable hand-ray. I had 2 separate projects for each plugin. Just follow the recommended settings.
8. I then built each one and deployed to the HL2 using these build settings:
Target Device: **HoloLens**
Architecture: **ARM64**
Build Type: **D3D Project**
Target SDK Version: **Latest installed** (which is 10.0.19041.0)
Minimum Platform Version: **10.0.10240.0**
Visual Studio Version: **Latest installed** (which is 2019)
Build and Run on: **USB Device**
Build Configuration: **Release**

## Expected behavior

I expected the hand-ray to be stable for both XR plugins, not just the OpenXR plugin.

## Screenshots

https://www.dropbox.com/sh/3n0v3lpnuc7snyt/AAAi4_RYTs4bvud4ERHNgMYXa?dl=0

Videos demonstrating the jittery hand-ray (WMR) and stable hand-ray (OpenXR). You will need to look closely as it may be hard to see the jittering. It's much more noticeable on the device.

## Your setup

- Unity Version 2020.3.13f1
- MRTK Version 2.7.2

## Target platform

- HoloLens 2

## Additional context

I could upload the Unity projects described above if necessary but they are pretty easy to create. This only occurs on the device, not in the Unity Editor.

**Comment by janezk7** [comment]
Was just about to make the same thread. 
Noticed that if you move the hand out of view and let it be detected again, it sort of "resets" and its smooth again. But only for another 5 seconds, then starts to jitter again. 
Most easily to see the jitter if you try to rotate just your wrist left and right (or up and down), while keeping the pointer in place.

**Comment by WikkidEdd** [comment]
We've got the same issue. We've also noticed that the general performance seems way worse when using WMR XR plugin vs OpenXR. For example in the HandExamples scene we get <30 fps when using WMR XR plugin.

**Comment by polar-kev** [comment]
Hey @elbuhofantasma, this is the issue I referenced at standup. Can you follow the repro steps and report back?

**Comment by polar-kev** [comment]
@ShayneRC, @janezk7 can you confirm that you saw this jitter while holding your hand still and that you were not rotating your hand at all?

**Comment by ShayneRC** [comment]
> @ShayneRC, @janezk7 can you confirm that you saw this jitter while holding your hand still and that you were not rotating your hand at all?

Yes this was happening while holding my hand still.

**Comment by polar-kev** [comment]
Thanks. @ShayneRC what are the specific ARR requirements you're referring to here?

**Comment by ShayneRC** [comment]
> Thanks. @ShayneRC what are the specific ARR requirements you're referring to here?

Using the OpenXR plugin causes our UI to be very wavy around the edges whereas the WMR plugin doesn't cause this.

**Comment by peterneil-bimholoview** [comment]
> We've got the same issue. We've also noticed that the general performance seems way worse when using WMR XR plugin vs OpenXR. For example in the HandExamples scene we get <30 fps when using WMR XR plugin.

Yes I can confirm (I work with Shayne).  Viewing the same asset on OpenXR vs WMR gives 60FPS vs 45FPS.

**Comment by elbuhofantasma** [comment]
I have followed the repro steps with Unity 2020.3.18f1 and found that the issue reproduces 100% of the time as shown in the videos. I am currently looking into the cause of the issue.

The jitter occurs at every distance possible when pointing the Hand Rays at objects, though it is significantly less noticeable at closer distances to an object or the edge of a Spatial Mapping boundary and can easily be mistaken for natural hand or arm movement. 

I reproduced this issue with the following configurations:

**Platform configuration**
_Unity 2020.3.18f1_
_Windows XR Plugin 4.5.0_
_UWP_

**Architecture**
_ARM_
_ARM64_

**Solution Configuration**
_Release_
_Debug_

**Comment by elbuhofantasma** [comment]
@ShayneRC Would you please file a separate issue about the wavy UI when using OpenXR?

**Comment by peterneil-bimholoview** [comment]
> @ShayneRC Would you please file a separate issue about the wavy UI when using OpenXR?

For our application the wavy UI is 100% expected for Depth LSR + ARR because we are dealing with a combination of near local rendered assets (UI) and far ARR rendered assets.   That is one of the reasons whey we use Planar LSR as it does not have that problem.

**Comment by peterneil-bimholoview** [comment]
Any update on this?  We have a App build that we cannot submit to Store until this jitter issue is resolved.

**Comment by polar-kev** [comment]
Hey @peterneil-bimholoview, no updates to share as yet.

**Comment by polar-kev** [comment]
@keveleigh, please post your updates here when you have them.

**Comment by keveleigh** [comment]
Update! I believe I've narrowed down the underlying issue in Windows XR Plugin and have proposed a fix. I'll work with Unity to get it released and will update the thread when it's out there!

**Comment by WikkidEdd** [comment]
That is good news. @keveleigh 

Does this fix both the hand jitter and the performance issues?

Do you have any time frame for this? We switched to OpenXR to not be blocked by this, but there are blocking issues in OpenXR as well.

Our current options are:
1. Wait for OpenXR Fixes 
2. Switch back to WMR XR Plugin when this fix comes out
3. Downgrade to Unity 2019 and Legacy XR

Any info you can share would help us make the right call.

**Comment by WikkidEdd** [comment]
For anyone else tracking this issue. The package with the jitter fix is available now 4.6.1

https://docs.unity3d.com/Packages/com.unity.xr.windowsmr@4.6/changelog/CHANGELOG.html

It doesn't mention addressing any of the performance issues in the release note, but I'm going to give it a go.

**Comment by keveleigh** [comment]
> It doesn't mention addressing any of the performance issues in the release note, but I'm going to give it a go.

I didn't do anything directly to address any performance issues you might be seeing, but the minor rework of source states I did _could_ help. Please let me know!

**Comment by WikkidEdd** [comment]
So far so good. The performance issues seem to be resolved too. 

There is a small issue which I had to fix up. I was getting this error every time a hand came in to view.

`Could not create a device for 'Hand - Left (XRInputV1)' (exception: System.ArgumentException: Expected control 'positionAccuracy' to be of type 'AxisControl' but is of type 'IntegerControl' instead!`

Inside the HololensHand class I had to change
`public AxisControl positionAccuracy { get; private set; }` to `public IntegerControl positionAccuracy { get; private set; }`
and
`positionAccuracy = GetChildControl<AxisControl>("positionAccuracy");` to `positionAccuracy = GetChildControl<IntegerControl>("positionAccuracy");`

**Comment by peterneil-bimholoview** [comment]
Been doing some testing on the 4.6.1 WMR version, but I am getting some really bad z-fighting type artifacts.  But only on some assets.  OpenXR and 4.5 WMR work fine

**Comment by keveleigh** [comment]
> Been doing some testing on the 4.6.1 WMR version, but I am getting some really bad z-fighting type artifacts. But only on some assets. OpenXR and 4.5 WMR work fine

Have you changed your Unity version as well? I'm not aware offhand of any rendering changes between 4.5 and 4.6.1 :/

**Comment by peterneil-bimholoview** [comment]
@keveleigh No still on 2020.3.13f

**Comment by keveleigh** [comment]
> So far so good. The performance issues seem to be resolved too.
> 
> There is a small issue which I had to fix up. I was getting this error every time a hand came in to view.
> 
> `Could not create a device for 'Hand - Left (XRInputV1)' (exception: System.ArgumentException: Expected control 'positionAccuracy' to be of type 'AxisControl' but is of type 'IntegerControl' instead!`
> 
> Inside the HololensHand class I had to change `public AxisControl positionAccuracy { get; private set; }` to `public IntegerControl positionAccuracy { get; private set; }` and `positionAccuracy = GetChildControl<AxisControl>("positionAccuracy");` to `positionAccuracy = GetChildControl<IntegerControl>("positionAccuracy");`

A fix for this was just merged into the plugin and should be in the next release! I'll update here when I have an ETA or when it's released (whichever comes first)

**Comment by polar-kev** [comment]
Update on the plugin release--we're anticipating that it will land on Jan 19th.

**Comment by polar-kev** [comment]
The Windows Mixed Reality Plugin updates are live. @ShayneRC, can you update and let us know if the issue has been fixed?

**Comment by peterneil-bimholoview** [comment]
@polar-kev  Yes the issue has been fixed in the WMR Plugin!

**Comment by chenritiana** [comment]
> underlying issue

Where is the HololensHand class? My package doesn't have this class.
